file/file_prefetch_buffer.cc

```diff
@@ -399,289 +376,285 @@ void FilePrefetchBuffer::ReadAheadSizeTuning(
 
   uint64_t roundup_len = end_offset - start_offset;
 
-  CalculateOffsetAndLen(alignment, start_offset, roundup_len, index, refit_tail,
+  CalculateOffsetAndLen(buf, alignment, start_offset, roundup_len, refit_tail,
                         chunk_len);
   assert(roundup_len >= chunk_len);
 
   // Update the buffer offset.
-  bufs_[index].offset_ = start_offset;
+  buf->offset_ = start_offset;
   // Update the initial end offset of this buffer which will be the starting
   // offset of next prefetch.
-  bufs_[index].initial_end_offset_ = initial_end_offset;
+  buf->initial_end_offset_ = initial_end_offset;
   read_len = static_cast<size_t>(roundup_len - chunk_len);
 
   UpdateReadAheadTrimmedStat((initial_end_offset - initial_start_offset),
                              (end_offset - start_offset));
 }
 
+// If data is overlapping between two buffers then during this call:
+//   - data from first buffer is copied into overlapping buffer,
+//   - first is removed from bufs_ and freed so that it can be used for async
+//     prefetching of further data.
 Status FilePrefetchBuffer::HandleOverlappingData(
     const IOOptions& opts, RandomAccessFileReader* reader, uint64_t offset,
-    size_t length, size_t readahead_size, bool& copy_to_third_buffer,
+    size_t length, size_t readahead_size, bool& copy_to_overlap_buffer,
     uint64_t& tmp_offset, size_t& tmp_length) {
+  // No Overlapping of data between 2 buffers.
+  if (IsBufferQueueEmpty() || NumBuffersAllocated() == 1) {
+    return Status::OK();
+  }
+
   Status s;
   size_t alignment = reader->file()->GetRequiredBufferAlignment();
-  uint32_t second;
+
+  BufferInfo* buf = GetFirstBuffer();
 
   // Check if the first buffer has the required offset and the async read is
   // still in progress. This should only happen if a prefetch was initiated
   // by Seek, but the next access is at another offset.
```

Comment:
Yes. So the data is overlapping between 2 buffers. Offset lies in first buffer so first buffer has the data but async is in progress for that buffer. So we call Poll here to get the data first and then copy the data to overlap_buf_.

We assumption was Poll is supposed to happen in `PrefetchInternal`. 
But for seek operation, we call `PrefetchAsync`  to submit async requests for all L0 files and L1-Ln levels. And then it calls again (second call) FilePrefetchBuffer to actually poll and read the data. It can be possible that second call wasn't made as data block was found in cache. So the buffer is never polled. 

This case handle that condition.
In short -  If buffer is needed but async is in progress, poll the data.

Reply:
> It can be possible that second call wasn't made as data block was found in cache.

The second call when there's a cache miss is like this, right? `SeekSecondPass()` -> ... -> `TryReadFromCache()` -> ... -> `PrefetchInternal()` 

I am seeing the code here poll even in that case, which is fine but seems to contradict the "next access is at another offset" part of the comment